site/content/blog/kapp-versioned-resource.md

```diff
@@ -0,0 +1,575 @@
+  4,  4   metadata:
+
+Changes
+
+Namespace  Name           Kind       Age  Op      Op st.  Wait to    Rs  Ri  
+default    simple-config  ConfigMap  3d   update  -       reconcile  ok  -  
+
+Op:      0 create, 0 delete, 1 update, 0 noop, 0 exists
+Wait to: 1 reconcile, 0 delete, 0 noop
+
+Continue? [yN]: y
+```
+
+Now will verify again the value for `MSG_KEY` in the `simple-app` pods.
+
+```bash
+$kubectl get pods            
+NAME                        READY   STATUS    RESTARTS   AGE
+simple-app-657f9c8494-t2pw9   1/1     Running   0          6m40s
+
+$kubectl exec -it simple-app-657f9c8494-t2pw9 sh
+# echo $MSG_KEY
+hello-carvel
+# exit
+#
+```
+Here, the value for env `MSG_KEY` is still not updated. To reflect the new changes of configmap we have to re-start the pod manually.
+
+```bash
+$kubectl delete pod simple-app-657f9c8494-t2pw9 
+pod "simple-app-657f9c8494-t2pw9" deleted
+
+$kubectl get pod 
+NAME                          READY   STATUS    RESTARTS   AGE
+simple-app-797ff748db-mqx97   1/1     Running   0          6s
+
+$kubectl exec -it simple-app-797ff748db-mqx97 sh
+# echo $MSG_KEY
+hello-carvel-india
+# 
+```
+After restarting the pod we can see the new changes we made in configmap.
+
+In above example, we saw that to reflect the changes of a configmap we need to restart the pod manually.
+
+## Versioned resource in `kapp`:
+
+kapp provides a [solution](https://carvel.dev/kapp/docs/v0.49.0/diff/#versioned-resources) for such scenarios, by offering a way to create uniquely named resources based on an original resource. Whenever we make a change to a resource marked as versioned, entirely new resource will get created by the kapp instead of updating the existing one. Also, it will update the new name to the referencing resource and re-start them to reflect the new changes.
+
+kapp has a concept of [versioned resources](https://carvel.dev/kapp/docs/v0.49.0/diff/#versioned-resources), where it creates a new version for resource whenever a change is made to it. To enable versioning we just need to add the annotation `kapp.k14s.io/versioned: ""` to the resource. Resources which are using this annotaion will follow the naming convention `{resource-name}-ver-{n}` where `n` will start with `1` and will get increamented by `1` after each update.
+
+Let's try to use this annotation for the ConfigMap from our previous example and see what happens when we make a change to it.
+
+```yaml
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: simple-config
+  annotations:
+    kapp.k14s.io/versioned: ""
+data:
+  hello_msg: hello-carvel
+
+---
+apiVersion: apps/v1
+kind: Deployment
+...
+```
+Now deploy the updated manifest and see the changes.
+
+```bash
+$kapp deploy -a app -f  app.yaml --diff-changes
+Target cluster 'https://127.0.0.1:33907' (nodes: minikube)
+
+@@ create configmap/simple-config-ver-1 (v1) namespace: default @@
+      0 + apiVersion: v1
+      1 + data:
+      2 +   hello_msg: hello-carvel-india
```

Comment:
Its not matching with the input. Input is `hello-carvel` instead of `hello-carvel-india`